fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -294,51 +365,123 @@ public FHIRRestOperationResponse doCreate(String type, Resource resource, String
             // Invoke the 'afterCreate' interceptor methods.
             getInterceptorMgr().fireAfterCreateEvent(event);
 
-            // Commit our transaction if we started one before.
+            // Commit the transaction if we started it (batch bundle)
             txn.commit();
             txn = null;
 
             return ior;
         } finally {
-            // Restore the original request context.
-            FHIRRequestContext.set(requestContext);
-
-            // If we previously started a transaction and it's still active, we need to rollback due to an error.
             if (txn != null) {
                 txn.rollback();
             }
-
-            log.exiting(this.getClass().getName(), "doCreate");
+            // Restore the original request context.
+            FHIRRequestContext.set(requestContext);
+            log.exiting(this.getClass().getName(), "doCreatePersist");
         }
     }
 
     @Override
     public FHIRRestOperationResponse doPatch(String type, String id, FHIRPatch patch, String ifMatchValue,
             String searchQueryString, boolean skippableUpdate) throws Exception {
+        log.entering(this.getClass().getName(), "doPatch");
 
         // Validate that interaction is allowed for given resource type
         validateInteraction(Interaction.PATCH, type);
+        
+        FHIRTransactionHelper txn = new FHIRTransactionHelper(getTransaction());
+        txn.begin();
+        
+        // Save the current request context.
+        FHIRRequestContext requestContext = FHIRRequestContext.get();
+        try {
+            // Fetch and patch the resource
+            List<Issue> warnings = new ArrayList<>();
+            FHIRRestOperationResponse metaResponse = doUpdateMeta(type, id, patch, null, ifMatchValue, searchQueryString, skippableUpdate, DO_VALIDATION, warnings);
+            if (metaResponse.isCompleted()) {
+                // skip the actual update because it would be a NOP
+                txn.commit();
+                txn = null;
+                return metaResponse;
+            }
```

Comment:
looks like this code is common between doPatch() and doUpdate() - perhaps could be moved into doPatchOrUpdate()